main.py

```diff
@@ -134,40 +142,45 @@ async def on_transaction(self, peer: Peer, payload: Transaction) -> None:
                 self.ez_send(peer, payload)
 
     @lazy_wrapper(BlockMessage)
-    async def on_block(self, peer: Peer, payload: BlockMessage) -> None:
-        unserialized_block = pickle.loads(payload.block)
-        print(unserialized_block)
+    async def receive_block(self, peer: Peer, payload: BlockMessage) -> None:
+        print('----------on block----------')
+        print(payload)
+
+        print(payload.block.previous_hash)
+        print(payload.block.merkle_hash)
+        print(payload.block.transactions)
 
         # Check if the block is already in the chain
-        if payload.hash not in [block.merkle_tree.get_root_hash() for block in self.blocks]:
-            print(f'Block {payload.hash} not in my chain {self.get_peer_id(peer)}')
-            logging.info(f'Block {payload.hash} not in my chain {self.get_peer_id(peer)}')
-            self.blocks.append(unserialized_block)
+        # if payload.hash not in [block.get_merkle_hash() for block in self.blocks]:
+        #     print(f'Block {payload.hash} not in my chain {self.get_peer_id(peer)}')
+        #     logging.info(f'Block {payload.hash} not in my chain {self.get_peer_id(peer)}')
+        #     self.blocks.append(payload.block)
 
-        # WIP: Necessary check?
-        # Check if the previous block hash is indeed the past block
+        # Remove block transactions from my pending_txs list
+        # for tx in unserialized_block.transactions:
+        #     if (tx.sender, tx.nonce) in [(tx.sender, tx.nonce) for tx in self.pending_txs]:
+        #         self.pending_txs.remove(tx)
 
     def finalize_and_broadcast_block(self):
-        self.current_block.merkle_tree.recalculate_tree()
-        new_block_hash = self.current_block.merkle_tree.get_root_hash()
+        self.current_block.update_tree()
+        new_block_hash = self.current_block.get_merkle_hash()
         print(f'New block hash: {new_block_hash}')
         logging.info(f'New block hash: {new_block_hash}')
         self.blocks.append(self.current_block)
-        serialized_block = pickle.dumps(self.current_block)
-        self.current_block = Block(new_block_hash)
 
-        self.broadcast_new_block(new_block_hash, serialized_block)
+        self.broadcast_block(new_block_hash, self.current_block)
+        self.current_block = Block(new_block_hash)
```

Comment:
here also "Expected type 'str', got 'bytes' instead"

Reply:
thanks, now it should be good